### ISSUE TITLE
Improve crypto verification?

### ISSUE THREAD

**Issue by ioerror** [mumble-voip/mumble#1811]
At the moment on the GNU/Linux client, I see a small dialog under server info with details. The "Control channel" option tells me that my connection is "Encrypted with 256 bit AES256-SHA" but it lacks other information. This is useful but it leaves me wondering: What is the key agreement protocol? Is my connection forward secret?

**Comment by hacst** [comment]
We output only the cipher name there currently. I guess we could add the key exchange method too. Unless you compiled against a recent Qt 5 with up to date OpenSSL you won't have PFS on Linux. On Windows and Mac we control the versions of these libraries so there you'll get it out of the box when using current snapshots.

**Comment by ioerror** [comment]
I'd like to support outputting the keyex method and all connection details - was it TLS 1.0? Was it TLS 1.2?

I'm using Debian 8 and that means that the program was compiled with QT 4.8.6 - ruh roh! I guess I'll ask the admin to rotate the certs and destroy the private keys.

If the server only allows forward secret modes of operation, I guess having this visible in the client would be a nice thing.

Any idea about other clients? I assume the iOS and Android clients are similarly problematic?

**Comment by mkrautz** [comment]
Yes, right now we're in the unfortunate situation that our 1.2 releases of Mumble on Linux are bound to Qt 4.

Qt 4 only exposes TLS 1.0. And Qt 4 doesn't allow DHE for QSsl-based servers. So that's the world Mumble 1.2 releases for Linux have to abide by. And that means no PFS, unfortunately.

Qt 5 does TLS 1.2, as well as ECDHE. It also does DHE, but with a static (non-changable by the programmer) 1024-bit DH group. I have a changeset in progress (https://codereview.qt-project.org/#/c/113070), which didn't make it into the Qt 5.6 release cycle, so it's probably going to be in Qt 5.7. But only Mumble 1.3 supports linking against Qt 5, and it's not released yet.

Sooo... Qt 4 is pretty bad in that regard.

Our own binary releases of Mumble 1.2 are based on a patched version of Qt 4, that does allow TLS 1.2 and ECDHE. Since Qt doesn't yet (pending on my patch) allow server admins to change DH params, we've opted to only allow ECDHE cipher suites in the default cipher suites for the client and for the server. But the default cipher suites can be overridden for both the server and clients. (See http://blog.mumble.info/mumble-1-2-9/)

Our own binary releases are: Windows client and server, OS X client and server, and Linux server. They are available from the front page of the website: http://wiki.mumble.info/wiki/Main_Page

Mumble 1.2 releases distributed by distros are not going to have the patched versions of Qt 4, so they're unfortunately stuck on TLS 1.0 without DHE, so no PFS. Both servers, and clients.

The iOS client is also TLS 1.0-only right now, but I hope to remedy that soon.

I do not know about Plumble.

So, to sum it up, if you want PFS on Linux, you're best of:

 - Running a server off git master, built against Qt 5.  Or run the static binary we provide via mumble.info (Both 1.2 and 1.3 will do TLS 1.2 with ECDHE/AES-GCM)

 - Running 1.3 clients built off git master, against Qt 5.

(Unless you patch Qt 4 to allow applications that ask for "TLS 1.0" to also allow greater versions...)

So, it's a pretty unfortunate situation. Our hope is to get Mumble 1.3 out of the door before the new year. But obviously, that'll not get into existing distros.

I also expect we'll do an update to the iOS app that allows it to negotiate TLS > 1.0 before the new year.

**Comment by ioerror** [comment]
Wowza - thank you for the detailed explaination!

I agree that this is extremely bad news all around - what happens when the server is running off of git master and has a mix of clients? Will any of them select the forward secret ECDHE methods at the moment? I guess no?

At least in theory, TLS 1.0 does offer some forward secret modes - so if a server only supported forward secret modes of operation, I could imagine that we might see some useful overlap, no?

I looked at the mumble client on Debian GNU/Linux and found that the TLS Client Hello offered TLS_ECDHE_RSA_* as well as other TLS_DHE_* cipher suites.

Wouldn't an upgraded server that disabled all other non-forward-secure suites overlap with the forward secret set on that client? Or is this only part of the problem and QT actually stops this from happening regardless of the TLS client hello cipher suite list?

**Comment by mkrautz** [comment]
If you're running a git master server, 1.2.9+ clients that run our patched Qt (our binary clients on Windows, OS X), as well as 1.3.x clients built against Qt 5, both ends will have the follow cipher preference by default (in OpenSSL notitation):

EECDH+AESGCM:AES256-SHA:AES128-SHA

So, in practice, those clients will all get TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384.

The last two entries yield TLS_RSA_WITH_AES_256_CBC_SHA and TLS_RSA_WITH_AES_128_CBC_SHA. They are there for backwards compatibility with older clients.
Server admins are free to remove them, if they wish, of course.

If you connect to the same server with a 1.2.x client from, say, Debian. I.e., one that does not have a patched Qt 4, it'll negotiate the CBC mode ciphers.

Regarding PFS in TLS 1.0, I believe the only cipher suites that grant you that are the TLS_DHE_*.

Qt 4 doesn't support DHE for the server-side, but Qt 5 does. However, it's a static, hardcoded 1024-bit DH group (2nd Oakley Group), and not changable via the Qt API at present. The changeset in Qt's gerrit I linked in my previous post is my implementation of an API that allows you to set custom DH params, however, it's not in Qt yet.

Because of the static DH group in Qt 5, we decided to only prefer ECDHE and not DHE in both client and server due to Logjam, etc. -- we'd like to have the server have settable DH params before we go all in. With that said, we should probably allow DHE on the client side to future proof ourselves, so that current clients will be able to connect to a "modern" Murmur instance that allows DHE.

If we do that, we should be able to have current 1.2.0 clients that only do TLS 1.0 negotiate the following PFS cipher suites:

    TLS_DHE_RSA_WITH_AES_128_CBC_SHA        DHE-RSA-AES128-SHA
    TLS_DHE_RSA_WITH_AES_256_CBC_SHA        DHE-RSA-AES256-SHA

That would make the story much better for Linux users that use distro packages of Mumble.

**Comment by ioerror** [comment]
On 9/16/15, Mikkel Krautz <notifications@github.com> wrote:
> If you're running a git master server, 1.2.9+ clients that run our patched
> Qt (our binary clients on Windows, OS X), as well as 1.3.x clients built
> against Qt 5, both ends will have the follow cipher preference by default
> (in OpenSSL notitation):
>
> EECDH+AESGCM:AES256-SHA:AES128-SHA

OK - that is good - that means that having an upgraded modern server
will greatly improve what is possible, which is expected...

>
> So, in practice, those clients will all get
> TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384.

I guess that this is OK. It isn't a djb curve but it will at least
give us forward secrecy after a connection has ended.

>
> The last two entries yield TLS_RSA_WITH_AES_256_CBC_SHA and
> TLS_RSA_WITH_AES_128_CBC_SHA. They are there for backwards compatibility
> with older clients.
> Server admins are free to remove them, if they wish, of course.

Is there an obvious way to do this with some configuration? Or is this
a compile time option?

>
> If you connect to the same server with a 1.2.x client from, say, Debian.
> I.e., one that does not have a patched Qt 4, it'll negotiate the CBC mode
> ciphers.
>
>
> Regarding PFS in TLS 1.0, I believe the only cipher suites that grant you
> that are the TLS_DHE_*.
>
> Qt 4 doesn't support DHE for the server-side, but Qt 5 does. However, it's a
> static, hardcoded 1024-bit DH group (2nd Oakley Group), and not changable
> via the Qt API at present. The changeset in Qt's gerrit I linked in my
> previous post is my implementation of an API that allows you to set custom
> DH params, however, it's not in Qt yet.

Ah - that is a problem. I wonder which is worse - no forward secrecy
with an easy to steal key or a  1024-bit DH group? Two bad choices.
:-(

>
> Because of the static DH group in Qt 5, we decided to only prefer ECDHE and
> not DHE in both client and server due to Logjam, etc. -- we'd like to have
> the server have settable DH params before we go all in. With that said, we
> should probably allow DHE on the client side to future proof ourselves, so
> that current clients will be able to connect to a "modern" Murmur instance
> that allows DHE.
>

Sure - that seems reasonable. I've also asked Tor's resident OpenSSL
and crypto expert to chime in here.

> If we do that, we should be able to have current 1.2.0 clients that only do
> TLS 1.0 negotiate the following PFS cipher suites:
>
>     TLS_DHE_RSA_WITH_AES_128_CBC_SHA        DHE-RSA-AES128-SHA
>     TLS_DHE_RSA_WITH_AES_256_CBC_SHA        DHE-RSA-AES256-SHA
>

That would be *fantastic* because the clients are already deployed.

> That would make the story much better for Linux users that use distro
> packages of Mumble.

Yes, I fully agree. How do we make this happen?

**Comment by mkrautz** [comment]
You can change the ciphers in murmur.ini by setting sslCiphers= to an OpenSSL cipher suite list.

In the client, there is a hidden (no UI for it) config option called "net/sslciphers" that can be set in the registry (Windows), preferences plist (OS X) or .ini file (*nix). It also takes an OpenSSL cipher suite.

Note however that these options were introduced in version 1.2.9. Previous versions use OpenSSL's default, which I believe can be affected by openssl.cnf...

**Comment by mkrautz** [comment]
As for enabling DHE support for Murmur in master, it's somehing I'm actively working on...  I am still not sure if the TLS 1.0 clients will be able to negotiate DHE.

**Comment by nmathewson** [comment]
Hi, I'm the resident openssl/crypto guy that ioerror mentioned above. :)  

This plan sounds pretty solid, if I understand it correctly; getting server-side support for good (large group) DHE in the meantime while more clients migrate to ECDHE would be excellent.

**Comment by mkrautz** [comment]
I have confirmed the ability to negotiate the TLS 1.0 EDH ciphersuites with clients that use Qt 4. I tested a Mumble 1.2.8 on Windows 8.1, and Mumble 1.2.8 on Ubuntu 14.10).

There's a PR, pending review, for adding support for setting custom DH params:
https://github.com/mumble-voip/mumble/pull/1816

**Comment by micah** [comment]
> I have confirmed the ability to negotiate the TLS 1.0 EDH ciphersuites with clients that use Qt 4. I tested a Mumble 1.2.8 on Windows 8.1, and Mumble 1.2.8 on Ubuntu 14.10).

I haven't been able to negotiate EDH ciphersuites between 1.2.10 server
and 1.2.10 client, and I'm a little puzzled about why.

I've tried a few different configs:

When I use the default one:
sslCiphers=EECDH+AESGCM:AES256-SHA:AES128-SHA - and connect, I see in
the terminal the following:

ServerHandler: TLS cipher preference is "ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:AES256-SHA:AES128-SHA"

but when I go to Server->Information, I see:

Version
Protocol 1.2.10
...
Control channel
Encrypted with 256 bit AES256-SHA
...
Voice channel
Encrypted with 128 bit OCB-AES128

any idea why that is?

**Comment by mkrautz** [comment]
@micah As discussed in this thread, Qt 4 only supports TLS 1.0, and doesn't allow you to set Diffie Hellman parameters for servers. So if you're testing 1.2.10 packages from a distro, you won't get anything other than you're seeing.

The comment you quote was about being able to get EDH ciphersuites negotiated on 1.2.x clients on Linux distros, when connecting to a server patched with this PR: https://github.com/mumble-voip/mumble/pull/1816
